src/mscorlib/src/System/Globalization/EncodingTable.Unix.cs

```diff
@@ -1,101 +1,162 @@
+            throw new ArgumentException(
+                String.Format(
+                    CultureInfo.CurrentCulture,
+                    Environment.GetResourceString("Argument_EncodingNotSupported"), name), "name");
         }
-    
-        [System.Security.SecuritySafeCritical]  // auto-generated
-        unsafe internal static CodePageDataItem GetCodePageDataItem(int codepage) {
-            // TODO: Implement this fully.
-            switch (codepage)
-            {
-                case CodePageUtf7:
-                    return new CodePageDataItem("utf-7", CodePageUtf7, "utf-7", "utf-7", 771);
 
-                case CodePageUtf8:
-                    return new CodePageDataItem("utf-8", CodePageUtf8, "utf-8", "utf-8", 771);
+        internal static CodePageDataItem GetCodePageDataItem(int codepage)
+        {
+            CodePageDataItem dataItem;
 
-                case CodePageUtf16:
-                    return new CodePageDataItem("utf-16", CodePageUtf16, "utf-16", "utf-16", 771);
+            // We synchronize around dictionary gets/sets. There's still a possibility that two threads
+            // will create a CodePageDataItem and the second will clobber the first in the dictionary. 
+            // However, that's acceptable because the contents are correct and we make no guarantees
+            // other than that. 
 
-                case CodePageUtf16BE:
-                    return new CodePageDataItem("utf-16BE", CodePageUtf16BE, "utf-16BE", "utf-16BE", 771);
+            //Look up the item in the hashtable.
+            dataItem = (CodePageDataItem)hashByCodePage[codepage];
 
-                case CodePageUtf32:
-                    return new CodePageDataItem("utf-32", CodePageUtf32, "utf-32", "utf-32", 771);
+            //If we found it, return it.
+            if (dataItem != null)
+            {
+                return dataItem;
+            }
 
-                default:
-                    return new CodePageDataItem("utf-8", CodePageUtf8, "utf-8", "utf-8", 771);
+            //If we didn't find it, try looking it up now.
+            //If we find it, add it to the hashtable.
+            //This is a linear search, but we probably won't be doing it very often.
+            for (int i = 0; i < encodingDataTableItems.Length; i++)
+            {
+                CodePageDataItem temp = encodingDataTableItems[i];
+                if (temp.CodePage == codepage)
+                {
+                    dataItem = temp;
+                    hashByCodePage[codepage] = dataItem;
+                    break;
+                }
             }
+
+            return dataItem;
         }
 
         // PAL ends here.
 
-        const int CodePageUtf7 = 65000;
-        const int CodePageUtf8 = 65001;
-        const int CodePageUtf16 = 1200;
-        const int CodePageUtf16BE = 1201;
-        const int CodePageUtf32 = 12000;
+        private static Dictionary<string, ushort> encodingDataTable = 
```

Comment:
Also, maybe specify the capacity based on the number of pairs being added, and add a `Debug.Assert` to ensure the capacity and count match? Along the lines of: https://github.com/dotnet/corefx/commit/425d8e7979fbefb7cdb5f44f557932ff5a9ae570